components/crud-web-apps/jupyter/backend/apps/common/form.py

```diff
@@ -135,6 +135,25 @@ def set_notebook_image_pull_policy(notebook, body, defaults):
     )
 
 
+def set_notebook_http_rewrite_uri(notebook, body, defaults):
+    notebook_annotations = notebook["metadata"]["annotations"]
+    http_rewrite_uri = get_form_value(body, defaults, "httpRewriteURI")
+    if http_rewrite_uri:
+        notebook_annotations["notebooks.kubeflow.org/http-rewrite-uri"] = http_rewrite_uri
+
+
+def set_http_headers_request_set(notebook, body, defaults):
+    notebook_annotations = notebook["metadata"]["annotations"]
+    http_headers_request_set = get_form_value(body, defaults, "httpHeadersRequestSet")
+    if http_headers_request_set:
+        notebook_annotations["notebooks.kubeflow.org/http-headers-request-set"] = http_headers_request_set
+
+
+def set_server_type(notebook, body, defaults):
+    notebook_annotations = notebook["metadata"]["annotations"]
+    notebook_annotations["notebooks.kubeflow.org/server-type"] = get_form_value(body, defaults, "serverType")
```

Comment:
Lets also add some logic here that would handle incorrect values.

The expected values here, from the request, would be:
* `None`, which means `"jupyter"`
* `"jupyter"`
* `"rstudio"`
* `"vscode"`

If this field is present in the request but has a different value from the expected ones then the backend should raise a `400` error

Reply:
Does adding error logic make sense here? The frontend doesn't rely on user input so unless there is a MITM change to the request a wrong input cannot happen. Also `none` cannot happen, as one of the toggles is always selected. It's not as if the JWA backend is decoupled from the frontend (they exist in the same container and are built together). 
Related to this, how would I go about causing a 400 error specifically in the backend code?

Reply:
Indeed, the frontend sends only specific values but it's a good practice for the backend to make it explicit to the user/frontend that the data it received was not formatted as expected.

Reply:
Should I handle the error with `log.error`?

Reply:
And you could look at this snippet for how to raise this error

https://github.com/kubeflow/kubeflow/blob/master/components/crud-web-apps/jupyter/backend/apps/common/form.py#L31-L34

Reply:
Thanks for the link. Sorry for wasting your time on this, I should have looked at the code a bit better and I would have found it.

Reply:
You are not wasting anyone's time. Your efforts are as important as mine and no releases were delayed because of a question :)